src/deadline/client/_pid_utils.py

```diff
@@ -1,85 +1,148 @@
-                )
-            except psutil.NoSuchProcess:
-                # No such process exists with the process id, so we can delete the pid file
-                logger.echo(f"Process with pid {pid} is not running. Deleting pid file.")
-
-                f.close()  # Close the file before over-writing it
-
-                # Create a new pid file with the current process id
-                return _obtain_pid_lock_atomically(pid_file_full_path, logger)
+            print_function_callback(
+                f"Download pid lock file does not exist at {pid_file_full_path}"
+            )
+
+            can_obtain_pid_lock = True
+
+        # 3. Try to read existing pid file to verify process with the pid lock is active & hasn't ended pre-maturely
+        else:
+            with open(pid_file_full_path, "r") as f:
+                pid = f.read()
+                try:
+                    psutil.Process(int(pid))
+                    # Process with the pid exists, so we cannot obtain a lock
+                    print_function_callback(
+                        f"Unable to acquire pid lock as process with pid {pid} exists on {pid_file_full_path}"
+                    )
+                    raise RuntimeError(
+                        f"Unable to acquire pid lock as process with pid {pid} exists on {pid_file_full_path}"
+                    )
+                except psutil.NoSuchProcess:
+                    # No such process exists with the process id, so we can delete the pid file
+                    print_function_callback(
+                        f"Process with pid {pid} is not running. Deleting pid file."
+                    )
+
+                    f.close()  # Required for windows
+
+                    # Delete existing pid file as pid is not active
+                    # Once we've determined the process in pid file isn't running we do nothing except:
+                    # Closing file (required for windows) and deleting it
+                    # This will keep race conditions across concurrent processes trying this to be least possible without locking
+                    # TODO Try to minimize potential race conditions from concurrent processes trying to remove pid file
+                    os.remove(pid_file_full_path)
+
+                    can_obtain_pid_lock = True
 
     except Exception:
-        # We already checked the file exists before reading it.
-        # For any other unexpected exceptions, we should not delete the pid file and raise an error.
+        # For any other unexpected exceptions, we should raise an error.
         raise
 
+    # 4. If we've determined we can obtain lock for this process, we do so now.
+    # We handle concurrency for processes racing for this pid lock by using atomic primitives
+    # The pid lock is obtained successfully for one process and fails for the other process
+    # We use os.rename() for windows and os.link() for linux/macOS as atomic primitives
+    if can_obtain_pid_lock:
```

Comment:
Under what conditions is this value False? I think the function will have raised an exception in that case if I'm reading it correctly.

Reply:
In that case this check should not be here, and the `can_obtain_pid_lock` variable should be deleted. The code should follow either the pattern where it raises exceptions on errors and returns on success, or it should never raise exceptions and return the boolean. By mixing the patterns, it takes more effort for someone to read and understand the code.